pkg/cmd/experimental/buildchain/buildchain.go

```diff
@@ -88,215 +90,101 @@ func NewEdge(fullname, to string) *Edge {
-			namespaces = append(namespaces, project.Name)
+			o.namespaces = append(o.namespaces, project.Name)
 		}
-	}
-	if len(namespaces) == 0 {
-		namespaces = append(namespaces, namespace)
+		return nil
 	}
 
-	// Get all build configurations
-	buildConfigList := make([]buildapi.BuildConfig, 0)
-	for _, namespace := range namespaces {
-		cfgList, err := oc.BuildConfigs(namespace).List(labels.Everything(), fields.Everything())
+	namespace := cmdutil.GetFlagString(cmd, "namespace")
+	if len(namespace) == 0 {
+		namespace, _, err = f.DefaultNamespace()
 		if err != nil {
 			return err
 		}
-		buildConfigList = append(buildConfigList, cfgList.Items...)
 	}
+	o.namespace = namespace
 
-	// Parse user input and validate specified image stream
-	streams := make(map[string][]string)
-	if !all && len(args) != 0 {
-		name, specifiedTag, err := parseTag(args[0])
-		if err != nil {
-			return err
-		}
-
-		// Validate the specified image stream
-		is, err := oc.ImageStreams(namespace).Get(name)
-		if err != nil {
-			return err
-		}
-		stream := join(namespace, name)
-
-		// Validate specified tag
-		tags := make([]string, 0)
-		exists := false
-		for tag := range is.Status.Tags {
-			tags = append(tags, tag)
-			if specifiedTag == tag {
-				exists = true
-			}
-		}
-		if !exists && !allTags {
-			// The specified tag isn't a part of our image stream
-			return fmt.Errorf("no tag %s exists in %s", specifiedTag, stream)
-		} else if !allTags {
-			// Use only the specified tag
-			tags = []string{specifiedTag}
-		}
+	return nil
+}
 
-		// Set the specified stream as the only one to output dependencies for
-		streams[stream] = tags
-	} else {
-		streams = getStreams(buildConfigList)
+// Validate returns validation errors regarding build-chain
+func (o *BuildChainOptions) Validate(cmd *cobra.Command, args []string) error {
+	if len(args) > 1 ||
+		(len(args) == 1 && o.allNamespaces) ||
+		(len(args) == 0 && o.allTags) ||
+		(o.allNamespaces && o.allTags) ||
+		(o.allNamespaces && len(cmdutil.GetFlagString(cmd, "namespace")) != 0) {
+		return cmdutil.UsageError(cmd, "Must pass nothing, an image stream name:tag combination, or specify the --all flag")
 	}
+	return nil
+}
 
-	if len(streams) == 0 {
-		return fmt.Errorf("no ImageStream available for building its dependency tree")
-	}
+// RunBuildChain contains all the necessary functionality for the OpenShift
+// experimental build-chain command
+func (o *BuildChainOptions) RunBuildChain(f *clientcmd.Factory, cmd *cobra.Command, args []string) error {
+	d := describe.NewChainDescriber(o.oc, o.output)
 
-	output := cmdutil.GetFlagString(cmd, "output")
-	for stream, tags := range streams {
-		for _, tag := range tags {
-			glog.V(4).Infof("Checking dependencies of stream %s tag %s", stream, tag)
-			root, err := findStreamDeps(stream, tag, buildConfigList)
+	if o.allNamespaces {
+		for _, namespace := range o.namespaces {
+			desc, err := d.Describe(namespace, "")
```

Comment:
I don't think this is the output we want.  You want one graph with all the content to traverse so that cross project links resolve the way you expect.